### PR TITLE
fix: ignore unsupported function/action

### PR DESCRIPTION
**Pull request by jjtang1985**
Entity type is not supported as parameter of function/action import. Instead of throwing an error, the generator now uses log with warning messages and ignore such function/action import.

Also, when an action/function import fails to find action/function, they are also ignored instead of throwing errors.

### CODE REVIEW

File: test-resources/odata-service-specs/v4/API_TEST_SRV/API_TEST_SRV.edmx [hunk 1]
```diff
@@ -255,8 +263,17 @@
...
+        <ActionImport Name="TestActionImportUnsupportedEdmTypes" Action="API_TEST_SRV.TestActionImportUnsupportedEdmTypes">
+        </ActionImport>
+
+        <ActionImport Name="TestActionImportNoParameterEntityReturnType" Action="API_TEST_SRV.TestActionImportNoParameterEntityReturnType">
+        </ActionImport>
+
+        <ActionImport Name="TestActionImportEntityParameterNoReturnType" Action="API_TEST_SRV.TestActionImportEntityParameterNoReturnType">
+        </ActionImport>
+
+        <ActionImport Name="TestActionImportNoActionDefined" Action="API_TEST_SRV.TestActionImportNoActionDefined">
```
Comment (jjtang1985) [inline test-resources/odata-service-specs/v4/API_TEST_SRV/API_TEST_SRV.edmx:275]: e.g, this action import has action `API_TEST_SRV.TestActionImportNoActionDefined`, which is not defined.

File: packages/generator/src/edmx-to-vdm/v2/function-import.ts [hunk 2]
```diff
@@ -17,28 +18,36 @@ export function generateFunctionImportsV2(
...
+      const swaggerDefinition = swaggerDefinitionForFunctionImport(
+        f.Name,
+        httpMethod,
+        serviceMetadata.swagger
+      );
+
+      // TODO 1571
+      if (f.Parameter && hasUnsupportedParameterTypes(f.Parameter)) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v2/function-import.ts:33]: I think we should just filter the functionImports before the map by doing this, i. e.
```ts
return edmxFunctionImports
  .filter(f => f.Parameter && hasUnsupportedParameterTypes(f.Parameter))
  .map(f => ...
```

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 3]
```diff
@@ -34,34 +35,48 @@ export function generateFunctionImportsV4(
...
 
-    const httpMethod = 'get';
-    const swaggerDefinition = swaggerDefinitionForFunctionImport(
-      functionImport.Name,
-      httpMethod,
-      serviceMetadata.swagger
-    );
+      if (!edmxFunction) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:47]: Same here, filter before map

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 4]
```diff
@@ -34,34 +35,48 @@ export function generateFunctionImportsV4(
...
+        serviceMetadata.swagger
+      );
+
+      // TODO 1571
+      if (
+        edmxFunction.Parameter &&
+        hasUnsupportedParameterTypes(edmxFunction.Parameter)
+      ) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:62]: And here

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts [hunk 5]
```diff
@@ -30,31 +31,44 @@ export function generateActionImportsV4(
...
 
-    const httpMethod = 'post';
-    const swaggerDefinition = swaggerDefinitionForFunctionImport(
-      actionImport.Name,
-      httpMethod,
-      serviceMetadata.swagger
-    );
+      if (!edmxAction) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.ts:40]: And here.

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts [hunk 6]
```diff
@@ -30,31 +31,44 @@ export function generateActionImportsV4(
...
+        httpMethod,
+        serviceMetadata.swagger
+      );
+      // TODO 1571
+      if (
+        edmxAction.Parameter &&
+        hasUnsupportedParameterTypes(edmxAction.Parameter)
+      ) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.ts:54]: And here.

File: packages/generator/src/edmx-to-vdm/edmx-to-vdm-util.ts [hunk 7]
```diff
@@ -155,3 +155,16 @@ export function typesForCollection(
 
 export const propertyJsType = (type: string): string | undefined =>
   type.startsWith('Edm.') ? edmToTsType(type) : undefined;
+
+export function hasUnsupportedParameterTypes(parameters: EdmxParameter[]) {
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/edmx-to-vdm-util.ts:159]: The name of the import would actually be nice in the log message. Maybe we should pass the whole function / action import here?

Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/edmx-to-vdm-util.ts:159]:
```suggestion
export function hasUnsupportedParameterTypes(parameters: EdmxParameter[]): boolean {
```

File: packages/generator/src/edmx-to-vdm/edmx-to-vdm-util.ts [hunk 8]
```diff
@@ -155,3 +155,16 @@ export function typesForCollection(
...
+  const ret = parameters.find(p => {
+    if (!isEdmType(p.Type)) {
+      logger.warn(
+        `Only Edm type is supported as action/function parameter now, but detects ${p.Type}. Ignore the whole function/action import.`
+      );
+      return true;
+    }
+    return false;
+  });
+  return ret !== undefined;
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/edmx-to-vdm-util.ts:169]:
```suggestion
  const unsupportedParameters = parameters.filter(p => !isEdmType(p.Type));
  if (unsupportedParameters) {
      logger.warn(
        `Unsupported function or action import parameter types ${unsupportedParameters.map(p => p.Type).join(', ')} found. The SAP Cloud SDK currently only supports Edm types in parameters. Skipping code generation for function/action import.`
    );
  }
  return !!unsupportedParameters;
```

**Review by marikaner — CHANGES_REQUESTED** [review]

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 9]
```diff
@@ -10,12 +10,13 @@ import {
...
 import { VdmComplexType, VdmEntity, VdmFunctionImport } from '../../vdm-types';
 import { parseFunctionImportReturnTypes } from '../common/action-function-return-types';
+import { hasUnsupportedParameterTypes } from '../edmx-to-vdm-util';
 import { findActionFunctionByImportName } from './action-function-util';
 
 function findFunctionForFunctionImport(
   functions: EdmxFunction[],
   functionImport: EdmxFunctionImport
-): EdmxFunction {
+): EdmxFunction | undefined {
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:19]: Can it be undefined?

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 10]
```diff
@@ -34,34 +35,47 @@ export function generateFunctionImportsV4(
...
+        httpMethod,
+        serviceMetadata.swagger
+      );
+
+      // TODO 1571 remove when supporting entity type as parameter
+      if (
+        hasUnsupportedParameterTypes(edmxFunction.Name, edmxFunction.Parameter)
+      ) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:61]: Combine it with the top, forget my other comments on this.

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 11]
```diff
@@ -34,34 +35,47 @@ export function generateFunctionImportsV4(
...
+        returnType: parseFunctionImportReturnTypes(
+          edmxFunction.ReturnType?.Type,
+          entities,
+          complexTypes,
+          extractResponse,
+          serviceMetadata.edmx.oDataVersion
+        )
+      };
+    })
+    .filter(e => e !== undefined) as VdmFunctionImport[];
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:80]:
```suggestion
    .filter(e => e);
```

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 12]
```diff
@@ -34,34 +35,40 @@ export function generateFunctionImportsV4(
...
+  return functionImports
+    .map(functionImport => {
+      const edmxFunction = findFunctionForFunctionImport(
+        functions,
+        functionImport
+      );
+      // TODO 1571 remove when supporting entity type as parameter
+      if (!edmxFunction || hasUnsupportedParameterTypes(edmxFunction)) {
+        return undefined;
+      }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:47]: Sorry @jjtang1985, what I forgot in our discussion last week was, that the error handling (logging) is still in the find method which is the wrong place in my opinion and that this is here to stay (unlike the `hasUnsupportedParameterTypes` check only).

Consider the following:
```ts
type JoinedFunctionImportData = { functionImport: EdmxFunctionImport, function: EdmxFunction };

function joinFunctionImportData(functionImports: EdmxFunctionImport[], functions: EdmxFunction[]): JoinedFunctionImportData[] {
  const functionImportsWithoutFunctions = [];
  const joinedFunctionImportData = functionImports.reduce((joined, functionImport) => {
    const edmxFunction = findFunctionForFunctionImport(
      functions,
      functionImport
    );
    
    if(edmxFunction) {
      return [
        ...joined, {
          functionImport,
          function: edmxFunction
        }
      ];
    }
    functionImportsWithoutFunctions.push(functionImport.Name);
    return joined;
  }, []);

  if (functionImportsWithoutFunctions) {
    logger.warn(functionImportsWithoutFunctions);
  }
  return joinedFunctionImportData;
}

export function generateFunctionImportsV4(
  serviceMetadata: ServiceMetadata,
  entities: VdmEntity[],
  complexTypes: Omit<VdmComplexType, 'factoryName'>[],
  formatter: ServiceNameFormatter
): VdmFunctionImport[] {
  const functions = parseFunctions(serviceMetadata.edmx.root);
  const functionImports = parseFunctionImports(serviceMetadata.edmx.root);
  const joinedFunctionData = joinFunctionImportData(functionImports, functions);

  return joinedFunctionData
    // TODO 1571 remove when supporting entity type as parameter
    .filter(({ function: edmxFunction}) => !hasUnsupportedParameterTypes(edmxFunction))
    .map(({ functionImport, function: edmxFunction }) => {
    const httpMethod = 'get';
    const swaggerDefinition = swaggerDefinitionForFunctionImport(
      functionImport.Name,
      httpMethod,
      serviceMetadata.swagger
    );

    return {
      ...transformFunctionImportBase(
        functionImport,
        edmxFunction.Parameter,
        swaggerDefinition,
        formatter
      ),
      httpMethod,
      returnType: parseFunctionImportReturnTypes(
        edmxFunction.ReturnType?.Type,
        entities,
        complexTypes,
        extractResponse,
        serviceMetadata.edmx.oDataVersion
      )
    };
  });
}
```

**Review by marikaner — COMMENTED** [review]
Minor notes

**Comment by jjtang1985** [comment]
I did some performance tests by running the generator with sap business one edmx file. The whole generation takes ~10 min, where generating files takes ~1 min and transpiling ts takes ~9 min. Considering the parsing step is just a very initial stage, I would argue, there is no huge performance impact after this change.

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 13]
```diff
@@ -10,21 +11,65 @@ import {
...
+      return joined;
+    },
+    []
+  );
+
+  if (functionImportsWithoutFunctions) {
+    logger.warn(
+      `Unable to find a function import with name: ${functionImportsWithoutFunctions}, in the list of given functions: ${functions
+        .map(f => f.Name)
+        .join(',')}`
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:67]:
```suggestion
      `Could not find functions referenced by the following function imports. Skipping code generation: ${functionImportsWithoutFunctions
        .map(f => `${f.Name} => ${f.Function}`)
        .join(', \n')}`
```

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 14]
```diff
@@ -10,21 +11,65 @@ import {
...
+      if (edmxFunction) {
+        return [
+          ...joined,
+          {
+            functionImport,
+            function: edmxFunction
+          }
+        ];
+      }
+      functionImportsWithoutFunctions.push(functionImport.Name);
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:57]:
```suggestion
      functionImportsWithoutFunctions.push(functionImport);
```

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts [hunk 15]
```diff
@@ -10,21 +11,65 @@ import {
...
+interface JoinedFunctionImportData {
+  functionImport: EdmxFunctionImport;
+  function: EdmxFunction;
+}
+
+function joinFunctionImportData(
+  functionImports: EdmxFunctionImport[],
+  functions: EdmxFunction[]
+): JoinedFunctionImportData[] {
+  const functionImportsWithoutFunctions: string[] = [];
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/function-import.ts:40]:
```suggestion
  const functionImportsWithoutFunctions = [];
```

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts [hunk 16]
```diff
@@ -10,17 +11,62 @@ import { ServiceMetadata } from '../../edmx-parser/edmx-file-reader';
...
+interface JoinedActionImportData {
+  actionImport: EdmxActionImport;
+  action: EdmxAction;
+}
+
+function joinActionImportData(
+  actionImports: EdmxActionImport[],
+  actions: EdmxAction[]
+): JoinedActionImportData[] {
+  const actionImportsWithoutActions: string[] = [];
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.ts:40]:
```suggestion
  const actionImportsWithoutActions = [];
```

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts [hunk 17]
```diff
@@ -10,17 +11,62 @@ import { ServiceMetadata } from '../../edmx-parser/edmx-file-reader';
...
+      if (edmxAction) {
+        return [
+          ...joined,
+          {
+            actionImport,
+            action: edmxAction
+          }
+        ];
+      }
+      actionImportsWithoutActions.push(actionImport.Name);
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.ts:54]:
```suggestion
      actionImportsWithoutActions.push(actionImport);
```

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts [hunk 18]
```diff
@@ -10,17 +11,62 @@ import { ServiceMetadata } from '../../edmx-parser/edmx-file-reader';
...
+    []
+  );
+
+  if (actionImportsWithoutActions) {
+    logger.warn(
+      `Unable to find a action import with name: ${actionImportsWithoutActions}, in the list of given actions: ${actions
+        .map(action => action.Name)
+        .join(',')}`
+    );
+  }
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.ts:66]:
```suggestion
      `Could not find actions referenced by the following action imports. Skipping code generation: ${actionImportsWithoutActions
        .map(f => `${f.Name} => ${f.Action}`)
        .join(', \n')}`
```

**Review by marikaner — APPROVED** [review]
Looks good to me!
Minor: I think the error messages could be a bit clearer and the formatting is difficult to read in the logs if there are multiple incorrect actions/functions imports.

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts (hunk 15 above)
Comment (jjtang1985) [reply packages/generator/src/edmx-to-vdm/v4/function-import.ts:40]: it won't work due to `Argument of type 'string' is not assignable to parameter of type 'never'.`

File: packages/generator/src/edmx-to-vdm/v4/action-import.ts (hunk 16 above)
Comment (jjtang1985) [reply packages/generator/src/edmx-to-vdm/v4/action-import.ts:40]: same as above

File: packages/generator/src/edmx-to-vdm/v4/function-import.ts (hunk 15 above)
Comment (marikaner) [reply packages/generator/src/edmx-to-vdm/v4/function-import.ts]: Then probably `: EdmxFunctionImport[]` with my suggestions below.

**Pull request merged**